components/brave_shields/resources/panel/components/tree-list/index.tsx

```diff
@@ -44,44 +56,117 @@ function groupByOrigin (data: Url[]) {
+                                 ResourceState.Blocked)
+    getPanelBrowserAPI().dataHandler.allowScriptsOnce(origins)
+  }
+
+  const handleBlockAllScripts = () => {
+    const origins =
+      getScriptsOriginsWithState(props.resourcesList.allowedList,
+                                 ResourceState.AllowedOnce)
+    getPanelBrowserAPI().dataHandler.blockAllowedScripts(origins)
+  }
+
+  const handleBlockScript = (name: string, state: ResourceState) => {
+    getPanelBrowserAPI().dataHandler.blockAllowedScripts([new URL(name).origin])
+  }
+
+  const handleAllowScript = (name: string, state: ResourceState) => {
+    getPanelBrowserAPI().dataHandler.allowScriptsOnce([new URL(name).origin])
+  }
+
+  const isAllowedSectionVisible = props.type === ResourceType.Script
+  const permissionButtonHandler =
+    isAllowedSectionVisible ? handleAllowScript : undefined
   return (
     <S.Box>
       <S.HeaderBox>
         <S.SiteTitleBox>
           <S.FavIconBox>
-            <img key={siteBlockInfo?.faviconUrl.url} src={siteBlockInfo?.faviconUrl.url} />
+            <img src={siteBlockInfo?.faviconUrl.url} />
           </S.FavIconBox>
           <S.SiteTitle>{siteBlockInfo?.host}</S.SiteTitle>
         </S.SiteTitleBox>
-        <S.Grid>
-          <span>{props.totalBlockedCount}</span>
-          <span>{props.blockedCountTitle}</span>
-        </S.Grid>
       </S.HeaderBox>
-      <S.TreeBox>
+      <S.Scroller>
+        {isAllowedSectionVisible && (
         <div>
-          {[...mappedData.keys()].map((origin, idx) => {
+          <ScriptsInfo>
+            <span>{props.resourcesList.allowedList.length}</span>
+            <span>{props.totalAllowedTitle}</span>
+            <span>{<a href="#" onClick={handleBlockAllScripts}>
+                {getLocale('braveShieldsBlockScriptsAll')}
+              </a>
+            }</span>
+          </ScriptsInfo>
+          <ScriptsList>
+            {[...allowedScriptsByOrigin.keys()].map((origin, idx) => {
+              return (<TreeNode
+                key={origin}
+                host={origin}
+                type={props.type}
+                state={ResourceState.AllowedOnce}
+                onPermissionButtonClick={handleBlockScript}
+                permissionButtonTitle={getLocale('braveShieldsBlockScript')}
+                resourceList={allowedScriptsByOrigin.get(origin) ?? []}
+              />)
+            })}
+          </ScriptsList>
+        </div>
+        )}
+        <ScriptsInfo>
+          <span>{props.resourcesList.blockedList.length}</span>
+          <span>{props.totalBlockedTitle}</span>
+          {isAllowedSectionVisible && (<span>
+            {<a href="#" onClick={handleAllowAllScripts}>
+                {getLocale('braveShieldsAllowScriptsAll')}
+              </a>
+            }</span>
+          )}
+        </ScriptsInfo>
+        <ScriptsList>
+          {[...blockedScriptsByOrigin.keys()].map((origin, idx) => {
```

Comment:
I think `blockedScriptsByOrigin` is a `Map`, so it doesn't support map. In addition, `keys()` is returning an iterator, so `[...blockedScriptsByOrigin.keys()]` is seems pretty reasonable here.

Reply:
You could also do

```
[...blockedScriptsByOrigin].map(([origin], idx) => ...)
```

but I don't know if it's any better/worse

Reply:
I didn't realize that it was a Map instead of just an array. 

You can also use `Array.from()` which can make the code more readable and understandable. The spread operator can easily be overlooked among other code, and it can be unclear when reading the code later.